gameboy_opengl_web/src/lib.rs

```diff
@@ -33,8 +33,44 @@ struct EmulatorState {
     js_ctx: Value,
     busy: bool,
     audio_underrun: Option<usize>,
+    save: Box<dyn SaveFile>,
+}
+//Generic object for saving emulator state
+trait SaveFile {
+    fn load(&mut self) -> Option<Vec<u8>>;
+    fn save(&mut self, ram: Vec<u8>);
+}
+struct LocalStorageSaveFile {
+    save_name: String,
+}
+impl LocalStorageSaveFile {
+    pub fn new(name: String) -> LocalStorageSaveFile {
+        return LocalStorageSaveFile { save_name: name };
+    }
+}
+impl SaveFile for LocalStorageSaveFile {
+    fn load(&mut self) -> Option<Vec<u8>> {
+        if let Some(ram_str) = window().local_storage().get(&self.save_name) {
+            let chars: Vec<char> = ram_str.chars().collect();
+            let bytes: Vec<u8> = chars
+                .chunks(2)
+                .map(|chunk| {
+                    let byte: String = chunk.iter().collect();
+                    u8::from_str_radix(&byte, 16).unwrap()
+                })
+                .collect();
+            return Some(bytes);
+        }
+        None
+    }
+    fn save(&mut self, ram: Vec<u8>) {
+        let save_str: String = ram.iter().map(|x| format!("{:02X}", x)).collect();
```

Comment:
I tried doing saving by mapping the ram to a string like this originally, but the problem was that it was just too slow. Some games can have up to 2MB of RAM, so mapping every single byte to a string every frame slows the game down by a lot

Reply:
How did you solve the issue?

Reply:
Heh, my solution is a little hack-y, but has worked well so far. Basically, we maintain a String that is an exact copy of whatever is in the cartridge RAM and whenever it changes, update the String in-place.

This initializes the Ram String to the RAM contents before the emulation starts
```
  let ram_str = Rc::new(RefCell::new(
        ram.iter().map(|byte| format!("{:02x}", byte)).collect(),
    ));
```

This set's up an event listener for whenever the contents of the Cartridge RAM changes
```
 self.gameboy
            .set_ram_change_callback(Box::new(move |address, value| {
                if has_battery {
                    let byte_chars: Vec<char> = format!("{:02x}", value).chars().collect();
                    let (first, second) = (byte_chars[0] as u8, byte_chars[1] as u8);
                    // unsafe because we are updating the RAM String in-place
                    // but since we are updating it with valid UTF-8, the updated string will also be UTF-8
                    unsafe {
                        let mut ram_str_ref = ram_str.borrow_mut();
                        let bytes = ram_str_ref.as_bytes_mut();
                        bytes[address * 2] = first;
                        bytes[address * 2 + 1] = second;
                    }
                    *should_save_to_local.borrow_mut() = true;
                }
            }));
```

This takes the O(n) Vec<u8> to String mapping to a O(1) in-place update, which is a lot faster, but requires a bit more code and unsafe Rust.